- Submission number **182**

The tracks to which we submitted it are

- **SS4**: _High Performance Computing in Bioinformatics_
- **SS8**: _Integration of data, methods and tools in biosciences_
- **SS19**: _Deep analysis of deep-sequencing data: for the discovery of genes, transcription factors, proteins and biomarkers_

### Reviewer 1

- Overall evaluation: 1: (weak accept)
- Reviewer's confidence:  4: (high)
- Recommended for journal publication:  3: (fair)

> The short abstract, entitled: “Bio4j: bigger, faster, leaner” present a high-performance cloud-enabled graph-based bioinformatics data platform, which integrates most data available in UniProt KB (SwissProt + Trembl), Gene Ontology (GO), UniRef (50, 90, 100), RefSeq, NCBI taxonomy, and Expasy Enzyme DBs. Therefore, this is a very useful tool for the scientific community in the field of bioinformatics, because Bio4j is open source, available under the AGPLv3 license.
> 
> The main question is: what are the main advantages and disadvantages with other wellknown bioinformatics data platform?
> 
> What about the graphical interface for the user?
> 
> This short abstract present an interesting bioinformatics data platform, that as pointed out, is growing year by year, however, and despite of the interest that the contribution may have, the abstract is extremely short and therefore is very difficult to assess the impact and novelty of the approach. 
> 
> Because the bioinformatics data platform presented is of great interest and very relevant, I recommend authors to extend this short abstract to a full paper for the conference. It is possible to conditional accept the paper, but authors should submit a detailed final version in .pdf format using the template of the conference.

### Reviewer 2

- Overall evaluation: 0: (borderline paper)
- Reviewer's confidence:  4: (high)
- Recommended for journal publication:  2: (poor)

> In this abstract, authors presents their Bio4j platform, which provides an integrative and quick way to access biological data from several well-known as Uniprot, GO, RefSeq, etc. Specifically, authors show the main updates and implementation in this platform.
> 
> I think it is a great opportunity to make Bio4j known for attendants in this conference, even though this abstract is only focused on updates and improvements of this platform. However, in order to make this contribution more interesting for attendants, I think it is necessary that authors extend it to a full paper.
> 
> I encourage author to describe in detail in this full paper the novel modules they have incorporated to Bio4j (i.e. the Statika-based one or the additional layer for the domain model). It could also interesting to make a comparative study (I find essential, for example, to know how Bio4j improves other databases in terms of consuming time).
> 
> In summary, I find this platform of high interest for this conference but more information should be provided in order to be definitively accepted.

----

the modified version is in the file [bio4j.md](bio4j.md), and the text for the email is in [reply.md](reply.md).